### ISSUE TITLE
Created multiple objects if more than one autocomplete urls are used together.

### ISSUE THREAD

**Issue by shubhamdipt** [yourlabs/django-autocomplete-light#814]
In my form, I have more than one field in one form having autocomplete feature. In such a case when I create a new object, it creates POST request more than one time (actually n^2 where n is the number of autocomplete fields).

Reason: for each select2 widget used in the form, the js static files get embedded in the html.

**Comment by jpic** [comment]
> In my form, I have more than one field in one form having autocomplete feature. In such a case when I create a new object, it creates POST request more than one time (actually n^2 where n is the number of autocomplete fields).

I've heard about this, but I haven't been able to reproduce this so far, for example, in this example: http://dal-yourlabs.rhcloud.com/admin/select2_one_to_one/tmodel/add/

> Reason: for each select2 widget used in the form, the js static files get embedded in the html.

Not that I know of, the js static files get embedded from the {{ form.media }} call.

**Comment by shubhamdipt** [comment]
You need to create two different fields with different widgets such that the widget's css and js are rendered twice. Because with each creation, the css and js are rendered. However I tried them outside the admin. It might be that it behaves different for admin and other templates.

**Comment by jpic** [comment]
I'm sure you're not trying to troll me, but it seems like you've completely ignored my answer.

**Comment by shubhamdipt** [comment]
Thats fine. Its totally upto you if you want to solve this issue. I forked it and made it work for me by removing the js css and rendering it separately just once in my templates. If you allow I can create  a sample app to reproduce this issue. It might be that I am not explaining you well enough.

**Comment by jpic** [comment]
> Thats fine. Its totally upto you if you want to solve this issue.

You've got to be kidding me, I've been maintaining this app for 4 years for free.

> If you allow I can create a sample app to reproduce this issue. It might be that I am not explaining you well enough.

Thank you in advance for doing this.

**Comment by shubhamdipt** [comment]
So I have added an app where you could see the issue.. if you allow me to push I can push it as two separate branches: one with the issue and one with the solution

**Comment by jpic** [comment]
I'm sorry, but I couldn't reproduce your problem with your fork.

I have found no other active branch: https://github.com/shubhamdipt/django-autocomplete-light/branches/stale

All it looks like, from my point of view here, based on the information you've given me, is that you don't understand how form media rendering works: https://docs.djangoproject.com/en/1.10/topics/forms/media/

So, the only configuration in which your problem should happen is:

- you have two forms
- you're showing {{ form1.media }} and {{ form2.media }}
- you should try form_media = form1.media + form2.media and render {{ form_media }} instead.

**Comment by shubhamdipt** [comment]
Yup that worked ... Thanks ... if you want there were some issues with migration files and regarding TestModel when I was trying the TestProject app. I already created another sample app for two forms outside admin. I can push it to you. I think it will be nice to have as another bit complicated example.

**Comment by jpic** [comment]
Ok, I think we need to update the documentation about this, because I think
you're not the only one !!

Thanks a HEAP for driving me through this.

**Comment by shubhamdipt** [comment]
Hi .. just figured out something new. There is an interference with django crispy forms on using {% crispy form %} (media files get rendered directly at the widgets location). However if I use it like {{ form|crispy }}, {{ form_media }} then works fine.

**Comment by jpic** [comment]
> If one uses autocomplete and then {% crispy form %} : then media files get rendered directly where the form-fireld is used in the template.

Really ? They must be doing `{% for field in form.fields %}{{ field.media }} {{ field }}{% endfor %}`, this defeats the purpose of form media, so that should be fixed in `{% crispy %}` I think.... It turns out you're facing a pretty complex problem after all, we're really going to need a FAQ with this things because it must drive some users pretty crazy ^^

**Comment by jpic** [comment]
Could you try if this prevents duplicate initialization ?

```
--- a/src/dal/static/autocomplete_light/autocomplete.init.js
+++ b/src/dal/static/autocomplete_light/autocomplete.init.js
@@ -33,17 +33,17 @@ element was cloned with data - which should be the case.
         return '';
     }
 
-    var initialized = [];
+    if (document.yl.initialized === undefined) document.yl.initialized = [];
 
     function initialize() {
         var element = $(this).get(0);
 
-        if (initialized.indexOf(element) >= 0) {
+        if (document.yl.initialized.indexOf(element) >= 0) {
             return
         }
 
         $(this).trigger('autocompleteLightInitialize');
-        initialized.push(element);
+        document.yl.initialized.push(element);
     }
 
     $(document).ready(function() {
```

But I'm not sure I want this to work even if the user duplicates script loading, so that's why i never so wanted to "fix" it: if the user is loading the JS twice then it's problem ... Perhaps detecting this condition, ie. checking if the script has already been loaded is better.

**Comment by jpic** [comment]
Feedback ?

@apinsard wthere you aware of this ?

**Comment by apinsard** [comment]
Nope I was not aware of this.

My opinion is that it should be avoided to include a JS multiple times. This is handled by Django as far as possible. In such cases where Django doesn't handle it, I think it's the responsibility of the developer to ensure static files are loaded only once. Namely : `form_media = form1.media + form2.media`.

I'm not sure I understood the issue with crispy_forms, but I guess this should be fixed on there side. I don't believe DAL can do anything about it but handling multiple inclusion of its scripts. Which is only working around the core issue.

**Comment by apnewberry** [comment]
If it's a formset, `{{formset.media}}` works great.
